lightbeam/validate.py

```diff
@@ -108,33 +292,136 @@ def load_local_descriptors(self):
                 if value!="": return value
             elif isinstance(payload[k], list):
                 for i in range(0, len(payload[k])):
-                    value = self.has_invalid_descriptor_values(payload[k][i], local_descriptors, path+("." if path!="" else "")+k+"["+str(i)+"]")
+                    value = self.has_invalid_descriptor_values(payload[k][i], path+("." if path!="" else "")+k+"["+str(i)+"]")
                     if value!="": return value
             elif isinstance(payload[k], str) and k.endswith("Descriptor"):
                 namespace = payload[k].split("#")[0]
                 codeValue = payload[k].split("#")[1]
-                # check f it's a local descriptor:
-                matching_local_descriptors = list(filter(lambda descriptor: type(descriptor)==dict and descriptor.get("namespace", "")==namespace and descriptor.get("codeValue", "")==codeValue, local_descriptors))
+                # check if it's a local descriptor:
+                matching_local_descriptors = list(filter(lambda descriptor:
+                                                            type(descriptor)==dict
+                                                            and descriptor.get("namespace", "")==namespace
+                                                            and descriptor.get("codeValue", "")==codeValue
+                                                         , self.local_descriptors or []))
                 if len(matching_local_descriptors)>0: return ""
                 # check if it's a remote descriptor:
                 if not self.is_valid_descriptor_value(namespace, codeValue):
                     return payload[k] + f" is not a valid descriptor value for {k}" + (" (at " + path + ")" if path!="" else "")
         return ""
 
+    # Validates descriptor values for a single payload (returns an error message or empty string)
+    def has_invalid_references(self, payload, path=""):
+        for k in payload.keys():
+            if isinstance(payload[k], dict) and not k.endswith("Reference"):
+                value = self.has_invalid_references(payload[k], path+("." if path!="" else "")+k)
+                if value!="": return value
+            elif isinstance(payload[k], list):
+                for i in range(0, len(payload[k])):
+                    value = self.has_invalid_references(payload[k][i], path+("." if path!="" else "")+k+"["+str(i)+"]")
+                    if value!="": return value
+            elif isinstance(payload[k], dict) and k.endswith("Reference"):
+                is_valid_reference = False
+                original_endpoint = util.pluralize_endpoint(k.replace("Reference",""))
+
+                # this deals with the fact that an educationOrganizationReference may be to a school, LEA, etc.:
+                endpoints_to_check = self.EDFI_GENERICS_TO_RESOURCES_MAPPING.get(original_endpoint, [original_endpoint])
```

Comment:
This seems like the right approach, especially in the near term. Maybe there's a slick way to automate this fact-finding but I have no issue with Lightbeam having built-in knowledge of the Ed-Fi. And it definitely reduces code complexity for now.